src/backend/access/transam/varsup.c

```diff
@@ -408,6 +409,9 @@ GetNewExternalObjectId(void)
 	/*
 	 * must perform check on External Oid range on
 	 * initial access of NextExternalOid
+	 *
+	 * It's needed for upgrade scenario from old version
+	 * of HAWQ which doesn't support dedicated oid pool for HCatalog objects
 	 */
 	if (!IsExternalOidInitialized)
 	{
```

Comment:
Can refactor this 
if(master_highest_used_oid() < FirstExternalObjectId)
{
    ResetExternalObjectId;
}
else
{
    ereport(....);
}